### DISCUSSION TITLE
How do you call member functions on existing JavaScript objects?

### DISCUSSION THREAD

**Discussion by cataggar** [dotnet/runtime#88517] [Q&A] [answered]
Using https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/import-export-interop?view=aspnetcore-7.0, I am not sure how to call member functions on existing `JSObject` instances. For example, here is TypeScript code that I would like to do in C#:

``` ts
  for (var [name,node] of context.program.sourceFiles){
    console.log(name);
  }
```

I would like to do the above in C#. Here is a failed attempt
``` cs
        var program = context.GetPropertyAsJSObject("program");
        var sourceFiles = program.GetPropertyAsJSObject("sourceFiles");
        
        // How do I call 'entries'?
        // This does not work.
        var entries = MapPrototype.entries(sourceFiles);
```

Where I was trying to define `entries` with:
``` cs
public partial class MapPrototype {
    [JSImport("entries")]
    public static partial JSObject entries(JSObject self);
}
```
cc @pavelsavara

**Comment by pavelsavara** [comment]
OOP in not supported in Net7. You need to create another method in JS, wrap it as `JSImport` and pass object as argument

**Reply by cataggar** [reply]
What about Net8? I have a local build of this repo that I'm using.

**Reply by pavelsavara** [reply, marked as answer]
Not yet, maybe Net9 but that's open discussion. OOP over interop boundary is not good idea in my opinion. It leads to chatty interactions with lots of overhead, GC pressure and other problems. I advise to create higher level functionality methods and interop with that instead.

**Reply by cataggar** [reply]
Makes sense. I'll do that.

Given those drawbacks, it can still be useful. Here is how it is done in Rust with wasm_bindgen:
https://rustwasm.github.io/wasm-bindgen/reference/attributes/on-js-imports/method.html

**Comment by kumariaditi52** [comment]
#[wasm_bindgen(method)] lets Rust call JavaScript object methods while correctly handling the this context.

**Comment by kumariaditi52** [comment]
The comment is pointing out that even though using #[wasm_bindgen(method)] has drawbacks, it’s useful when you need Rust to call methods on JS objects.

**Comment by kumariaditi52** [comment]
The comment is pointing out that even though using #[wasm_bindgen(method)] has drawbacks, it’s useful when you need Rust to call methods on JS objects.

**Comment by kumariaditi52** [comment]
sourceFiles.InvokeVoid("forEach", (JSObject node, string name) =>
{
    Console.WriteLine(name);
});
